app/master/build_request_handler.py

```diff
@@ -73,6 +75,22 @@ def _build_preparation_loop(self):
         """
         while True:
             build = self._request_queue.get()
+            project_id = build.project_type.project_id()
+
+            if project_id not in self._project_preparation_locks:
+                self._project_preparation_locks[project_id] = Lock()
+
+            project_lock = self._project_preparation_locks[project_id]
+            async_preparation_thread = SafeThread(target=self._prepare_build_async, name='BuildPreparationThread',
+                                                  args=[build, project_lock])
+            async_preparation_thread.start()
+
+    def _prepare_build_async(self, build, project_lock):
+        """
+        :type build: Build
+        :type project_lock: Lock
+        """
```

Comment:
That's correct, the 'with project_lock:' is a short-hand for calling a project_lock.acquire() at the beginning and calling project_lock.release() at the end when the block exits.

When multiple threads call acquire, then I don't think there is any contractual guarantee on the order in which the threads receive the locks, no. It is an issue that we were aware of when designing it, but we decided that it was minor enough to not matter.